src/java.base/share/classes/sun/security/util/SecurityProviderConstants.java

```diff
@@ -105,13 +108,33 @@ public static final int getDefDSASubprimeSize(int primeSize) {
         }
     }
 
+    public static final int getDefAESKeySize() {
+        int currVal = DEF_AES_KEY_SIZE.get();
+        if (currVal == -1) {
+            int v = 256; // default AES key size
+            try {
+                // adjust if crypto policy only allows a smaller value
+                int max = Cipher.getMaxAllowedKeyLength("AES");
+                if (v > max)  {
+                    v = max;
+                }
+            } catch (NoSuchAlgorithmException ne) {
+                // should never happen; ignore and use the default
+            }
+            DEF_AES_KEY_SIZE.compareAndSet(-1, v);
+            currVal = v;
+        }
+        return currVal;
+    }
+
```

Comment:
I'm not very sure of this method.  Is it performance friendly if making the default key size calculation in the static block (from line 142 to the end of the file)?  Then, the DEF_AES_KEY_SIZE could be a public primitive int.

Or did I miss something?

Reply:
My very first prototype is to implement the AES keysize calculation as you commented, i.e. in the static block and use an int for DEF_AES_KEY_SIZE. However, it is later discovered through testing that this leads to deadlocks as this interferes with provider loading. Given that AES key size is just a small piece of the whole puzzle, it seems safer to defer this to a later point when it's actually needed rather than re-writing the code just to make it a static int. Performance-wise, this is a very small piece, generally should just be the AtomicInteger.get().

Reply:
I see.

Would you mind add a comment about the provider loading impact, just in case someone else have similar questions in the future?

Reply:
Sure, I can do that. Will add a comment about this.

Reply:
Thank you.  I have no more comment.